File: RETOS/Reto08/reto08.py
'''
 * Crea un programa que cuente cuantas veces se repite cada palabra
 * y que muestre el recuento final de todas ellas.
 * - Los signos de puntuación no forman parte de la palabra.
 * - Una palabra es la misma aunque aparezca en mayúsculas y minúsculas.
 * - No se pueden utilizar funciones propias del lenguaje que
 *   lo resuelvan automáticamente.
'''
import string

def contar_palabras(texto):
    # Convertir a minúsculas y eliminar signos de puntuación
    texto_limpio = texto.lower().translate(str.maketrans('', '', string.punctuation))

    # Dividir el texto en palabras
    palabras = texto_limpio.split()

    # Usar un diccionario para contar las palabras
    contador_palabras = {}

    for palabra in palabras:
        if palabra:
            if palabra in contador_palabras:
                contador_palabras[palabra] += 1
            else:
                contador_palabras[palabra] = 1

    return contador_palabras

# Leer el texto del usuario
texto = input("Ingresa el texto: ")

# Contar palabras y mostrar resultados
conteo = contar_palabras(texto)
for palabra, cantidad in conteo.items():
    print(f"{palabra}: {cantidad}")

File: RETOS/Reto08/reto08.kt
/*
 * Crea un programa que cuente cuantas veces se repite cada palabra
 * y que muestre el recuento final de todas ellas.
 * - Los signos de puntuación no forman parte de la palabra.
 * - Una palabra es la misma aunque aparezca en mayúsculas y minúsculas.
 * - No se pueden utilizar funciones propias del lenguaje que
 *   lo resuelvan automáticamente.
 */
import java.util.Scanner

fun contarPalabras(texto: String): Map<String, Int> {
    // Convertir a minúsculas y eliminar signos de puntuación
    val textoLimpio = texto.lowercase().replace(Regex("[^a-z\\s]"), "")

    // Dividir el texto en palabras
    val palabras = textoLimpio.split(Regex("\\s+"))

    // Usar un mapa para contar las palabras
    val contadorPalabras = mutableMapOf<String, Int>()

    palabras.forEach { palabra ->
        if (palabra.isNotEmpty()) {
            contadorPalabras[palabra] = contadorPalabras.getOrDefault(palabra, 0) + 1
        }
    }

    return contadorPalabras
}

fun main() {
    val scanner = Scanner(System.`in`)
    println("Ingresa el texto:")
    val texto = scanner.nextLine()

    val conteo = contarPalabras(texto)
    conteo.forEach { (palabra, cantidad) ->
        println("$palabra: $cantidad")
    }
}
